Android/live data convenience (#28)
MapsIndoors/docs: src/android/v3/live-data/index.md

        .setMultiLevelWildcard()
        .build();
liveDataManager.subscribeTopic(liveTopicCriteria);
        </code></pre>
    </mi-tab-panel>
    <mi-tab-panel id="kotlin">
        <h3>kotlin</h3>
        <pre lang ="Kotlin"><code>
var liveDataManager = LiveDataManager.getInstance();
liveDataManager.setOnLiveDataManagerStateChangedListener {
    Log.d("LiveDataState", "Live data manager state changed to: $it")
}
var liveTopicCriteria = LiveTopicCriteria.getBuilder("datasetId")
    .setMultiLevelWildcard()
    .build();
liveDataManager.subscribeTopic(liveTopicCriteria);
        </code></pre>
    </mi-tab-panel>
</mi-tabs>

In the example, the Topic is created using the `datasetId` and a multilevel wildcard, which will return all Live Data in the Solution.

## Rendering Live Data Locations

As mentioned `MapControl` has a default way of rendering Live Data Locations if you call `enableLiveData(String domainType)`. If you need to show live data in another way, you can add handlers for this either through `setOnWillUpdateLocationsOnMap(OnWillUpdateLocationsOnMap listener)` or calling `enableLiveData(String domainType, OnLiveLocationUpdateListener OnLiveLocationUpdateListener)`.

> Here are examples of using the different methods to render live data on your map:

Reviewer: ```suggestion
Here are examples of using the different methods to render Live Data on your map:
```

<mi-tabs>
    <mi-tab label="Java" tab-for="java"></mi-tab>
    <mi-tab label="Kotlin" tab-for="kotlin"></mi-tab>
    <mi-tab-panel id="java">

>         <h3>java</h3>

Reviewer: @timMikkelsen `java` lowercase or `Java`?

```suggestion
        <h3>Java</h3>
```

        <pre lang="Java"><code>
mMapControl.setOnWillUpdateLocationsOnMap(locations -> {
   for (MPLocation location : locations) {
       LiveUpdate occupancy = location.getLiveUpdate("occupancy");
       LocationDisplayRule currentDisplayRule = mMapControl.getDisplayRule(location);
       String displayRuleName = location.getId() + "_live";
       if (occupancy != null) {
           OccupancyProperty occupancyProperty = occupancy.getOccupancyProperties();
           LocationDisplayRule occupancyDisplayRule = new LocationDisplayRule
                   .Builder(displayRuleName, currentDisplayRule)
                   .setLabel("people = " + occupancyProperty.getNrOfPeople())
                   .build();
           mMapControl.setDisplayRule(occupancyDisplayRule, location);
       }
   }
});
        </code></pre>
    </mi-tab-panel>
    <mi-tab-panel id="kotlin">

>         <h3>kotlin</h3>

Reviewer: @timMikkelsen Same here; lowercase or uppercase?

```suggestion
        <h3>Kotlin</h3>
```

        <pre lang ="Kotlin"><code>
mMapControl.setOnWillUpdateLocationsOnMap { locations: List&lt;MPLocation&gt; ->
    for (location in locations) {
        val occupancy = location.getLiveUpdate("occupancy")
        val currentDisplayRule = mMapControl.getDisplayRule(location)
        val displayRuleName = location.id + "_live"
        if (occupancy != null) {
            val occupancyProperty = occupancy.occupancyProperties
            val occupancyDisplayRule =
                LocationDisplayRule.Builder(displayRuleName, currentDisplayRule!!)
                    .setLabel("people = " + occupancyProperty.nrOfPeople)
                    .build()
            mMapControl.setDisplayRule(occupancyDisplayRule, location)
        }
    }
}
        </code></pre>
    </mi-tab-panel>
</mi-tabs>

<mi-tabs>
    <mi-tab label="Java" tab-for="java"></mi-tab>
    <mi-tab label="Kotlin" tab-for="kotlin"></mi-tab>
    <mi-tab-panel id="java">

>         <h3>java</h3>

Reviewer: ```suggestion
        <h3>Java</h3>
```

        <pre lang="Java"><code>
mMapControl.enableLiveData(LiveDataDomainTypes.OCCUPANCY_DOMAIN, location -> {
   LiveUpdate occupancy = location.getLiveUpdate("occupancy");
   LocationDisplayRule currentDisplayRule = mMapControl.getDisplayRule(location);
   String displayRuleName = location.getId() + "_live";
   if (occupancy != null) {
       OccupancyProperty occupancyProperty = occupancy.getOccupancyProperties();
       LocationDisplayRule occupancyDisplayRule = new LocationDisplayRule
               .Builder(displayRuleName, currentDisplayRule)
               .setLabel("people = " + occupancyProperty.getNrOfPeople())
               .build();
       mMapControl.setDisplayRule(occupancyDisplayRule, location);
   }
});
        </code></pre>
    </mi-tab-panel>
    <mi-tab-panel id="kotlin">

>         <h3>kotlin</h3>

Reviewer: ```suggestion
        <h3>Kotlin</h3>
```

        <pre lang ="Kotlin"><code>
mMapControl.enableLiveData(LiveDataDomainTypes.OCCUPANCY_DOMAIN) { location: MPLocation ->
    val occupancy = location.getLiveUpdate("occupancy")
    val currentDisplayRule = mMapControl.getDisplayRule(location)
    val displayRuleName = location.id + "_live"
    if (occupancy != null) {
        val occupancyProperty = occupancy.occupancyProperties
        val occupancyDisplayRule =
            LocationDisplayRule.Builder(displayRuleName, currentDisplayRule!!)
                .setLabel("people = " + occupancyProperty.nrOfPeople)
                .build()
        mMapControl.setDisplayRule(occupancyDisplayRule, location)
    }
}
        </code></pre>
    </mi-tab-panel>
</mi-tabs>

> Note that since there is no guarantee of what Live Data you receive first and Locations can have multiple Live Data updates on different domains. It can be advised to check the lastModifiedTimeStamp of each live data update to choose what you want to render.

Reviewer: ```suggestion
Note that since there is no guarantee of which Live Data you receive first, and Locations can have multiple Live Data updates on different domains, we recommend checking the `lastModifiedTimeStamp` of each Live Data update to select which one to render.
```

## Handling Live Data Events

While only a few lines of code can get things moving around on a map, there are of course more handles that are relevant to create a robust and user-friendly real-time map experience.

### Listening for Live Updates

There are two ways to be notified about Live Updates:

1. On a general level through `OnReceivedLiveUpdateListener`, which is suitable in scenarios where all Live Updates might potentially affect the end user’s decisions, for example when searching broadly for an available meeting room.
2. On a map-specific level through `MapControl`, which is suitable in scenarios where the map is the context for the user’s actions, for example when browsing the map for an available meeting room nearby

To get Live Updates on a general level the `OnReceivedLiveUpdateListener` must be set on the `LiveDataManager`:

<mi-tabs>
    <mi-tab label="Java" tab-for="java"></mi-tab>
    <mi-tab label="Kotlin" tab-for="kotlin"></mi-tab>
    <mi-tab-panel id="java">

>         <h3>java</h3>

Reviewer: ```suggestion
        <h3>Java</h3>
```

        <pre lang="Java"><code>
liveDataManager.setOnReceivedLiveUpdateListener((topic, message) -> {
           if (message.getDomainType().equals(LiveDataDomainTypes.OCCUPANCY_DOMAIN)) {
               int people = message.getOccupancyProperties().getNrOfPeople();
               ...
           }
        });
        </code></pre>
    </mi-tab-panel>
    <mi-tab-panel id="kotlin">

>         <h3>kotlin</h3>

Reviewer: ```suggestion
        <h3>Kotlin</h3>
```

        <pre lang ="Kotlin"><code>
liveDataManager.setOnReceivedLiveUpdateListener { liveTopic, liveUpdate ->
            if (liveUpdate.domainType == LiveDataDomainTypes.OCCUPANCY_DOMAIN) {
                var people = liveUpdate.occupancyProperties.nrOfPeople
                ...
            }
        }
        </code></pre>
    </mi-tab-panel>
</mi-tabs>

To get Live Updates on a map-specific level, the `OnWillUpdateLocationsOnMap` must be set on `MapControl`:

<mi-tabs>
    <mi-tab label="Java" tab-for="java"></mi-tab>
    <mi-tab label="Kotlin" tab-for="kotlin"></mi-tab>
    <mi-tab-panel id="java">

>         <h3>java</h3>

Reviewer: ```suggestion
        <h3>Java</h3>
```

        <pre lang="Java"><code>
mapControl.setOnWillUpdateLocationsOnMap(locations -> {
    for(MPLocation location : locations){
        LiveUpdate liveUpdate = location.getLiveUpdate(LiveDataDomainTypes.OCCUPANCY_DOMAIN);
        ...
    }
}
        </code></pre>
    </mi-tab-panel>
    <mi-tab-panel id="kotlin">

>         <h3>kotlin</h3>

Reviewer: ```suggestion
        <h3>Kotlin</h3>
```
